### DISCUSSION TITLE
Edgex-Compose and latest docker

### DISCUSSION THREAD

**Discussion by sounak-1989** [edgexfoundry/community#49] [Q&A] [answered]
Hi guys,
I am using the Beaglebone AI-64 board which is an arm64 architecture. I cloned the edgex-compose repository and installed docker and docker-compose. Checked the versions as well to verify that the proper binaries for the architecture is included.

When I try to run the docker-compose file which I have done several times on my Linux PC and arm32 Beaglebone Black, this time I am getting an error from Docker saying
ERROR: In file './docker-compose.yml', service 'name' must be a mapping not a string.

I am not an expert in Docker but from what I saw on the internet this could be caused by indentations. I am using the command
"make gen no-secty arm64" to generate the yaml file so there should not be any indentation issues since it is auto generated. I have used this process several times before but did not get this error.
Below is the version of my docker
Client: Docker Engine - Community
 Version:           20.10.17
 API version:       1.41
 Go version:        go1.17.11
 Git commit:        100c701
 Built:             Mon Jun  6 23:02:34 2022
 OS/Arch:           linux/arm64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.17
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.17.11
  Git commit:       a89b842
  Built:            Mon Jun  6 23:01:01 2022
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.6.6
  GitCommit:        10c12954828e7c7c9b6e0ea9b0c02b01407d3ae1
 runc:
  Version:          1.1.2
  GitCommit:        v1.1.2-0-ga916309
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
docker-compose version 1.25.0, build unknown
docker-py version: 4.1.0
CPython version: 3.9.2
OpenSSL version: OpenSSL 1.1.1n  15 Mar 2022

Any help is much appreciated. Thanks

**Comment by bnevis-i** [comment]
Which version of the repo did you clone?  If you cloned the tip (main), then you no longer need to use docker-compose, but instead "docker compose" (the builtin).

**Reply by sounak-1989** [reply]
Yes I cloned the tip. I upgraded my docker to latest as well. 
Verified the Makefile. Seems like it is using "docker compose" notation. But still getting the error.

**Comment by sounak-1989** [comment, marked as answer]
Figured out the issue.  The latest docker compose for arm64 is 1.26 which does not support the "docker compose" notation. I reverted my edgex-compose back to Kamakura which uses "docker-compose" notation and it started working.

**Reply by bnevis-i** [reply]
> Figured out the issue. The latest docker compose for arm64 is 1.26 which does not support the "docker compose" notation. I reverted my edgex-compose back to Kamakura which uses "docker-compose" notation and it started working.

Ouch!  Are there plans to fix that?  Or are we going to need to revert the change?

**Reply by lenny-goodell** [reply]
Interesting that `docker compose` isn't available for arm64 when they have totally deprecated `docker-compose`. 

@sounak-1989 , please note that the Levski 2.3.0 release still supports use of `docker-compose`. The Levski compose builder will use `docker-compose` if it is installed. The switch to only `docker compose` is only in the main for the upcoming `Minnesota` release.

**Reply by bnevis-i** [reply]
@sounak-1989 

I cannot reproduce your issue.

I just spun up an arm64 vm, installed the latest docker engine, and have no issues.

![image](https://user-images.githubusercontent.com/21301128/222033045-f06b90a8-8e36-4d18-9fdb-fde02e586d5a.png)

Docker version 23.0.1, build a5ee5b1

Linux ubu2204-aarch64-edgex 5.15.0-60-generic #66-Ubuntu SMP Fri Jan 20 14:34:57
 UTC 2023 aarch64 aarch64 aarch64 GNU/Linux

**Reply by sounak-1989** [reply]
I am using Debian Bullseye on a Beaglebone AI-64. Is it possible that the Debian repository does not support the "docker compose" notation yet?

**Reply by bnevis-i** [reply]
Do not use the built-in packages.  They are ancient.  Please install docker directly using the instructions here:

https://docs.docker.com/engine/install/ubuntu/

Make sure you get the docker-ce / docker engine, not docker desktop.

**Reply by sounak-1989** [reply]
Yeah I went through the installation process on Docker website. But it was not installing the latest for some reason.

**Reply by bnevis-i** [reply]
I haven't tested Bullseye.  Give me a few days to get back to you.

**Reply by bnevis-i** [reply]
Bullseye also appears to work.  https://docs.docker.com/engine/install/debian/

**Reply by sounak-1989** [reply]
I tried it on my Beaglebone Black which is Bullseye but arm32. It seems to work. But for some reason it did not work on my Beaglebone AI-64 which is an arm64. Thanks for the help though.
